File: Spark_FastQ_BaseQual_Baseline/src/main/scala/FastQ_PerBaseQual_Mapper.scala
package main.scala

import main.scala.utils.AvgCount
import org.apache.hadoop.io.Text
import org.seqdoop.hadoop_bam.SequencedFragment

import scala.collection.mutable.ListBuffer

/**
  * master-thesis Clemens Banas
  * Organization: DBIS - University of Innsbruck
  * Created 21.04.2016
  */
object FastQ_PerBaseQual_Mapper {
  private val OFFSET: Int = 33

  def flatMap(sampleIdentifier:String, record: SequencedFragment): TraversableOnce[Pair[Pair[String, Int], Int]] = {
    val resList = new ListBuffer[Pair[Pair[String, Int], Int]]()
    val quality: Text = record.getQuality
    for(i <- 0 until quality.getLength) {
      resList.append(new Pair(new Pair(sampleIdentifier, (i+1)), getCorrespoindingIntValue(quality.charAt(i))))
    }
    resList.toTraversable
  }

  def map(pair: Pair[Pair[String, Int], AvgCount]): Pair[Pair[String, Int], Double] = {
    new Pair(pair._1, pair._2.getMeanValue())
  }

  private def getCorrespoindingIntValue(qualVal: Int): Int = {
    (qualVal - OFFSET)
  }

}

File: MapReduce_FastQ_BaseQual_SingleFile_Combiner/src/main/java/baseQual/FastQ_PerBaseQual_Mapper.java
package baseQual;

import org.apache.hadoop.io.IntWritable;
import org.apache.hadoop.io.Text;
import org.seqdoop.hadoop_bam.SequencedFragment;
import utils.QualityCountHelperWritable;

import java.io.IOException;

/**
 * master-thesis Clemens Banas
 * Organization: DBIS - University of Innsbruck
 * Created 21.04.2016
 */
public class FastQ_PerBaseQual_Mapper extends org.apache.hadoop.mapreduce.Mapper<Object, SequencedFragment, IntWritable, QualityCountHelperWritable> {
    private static final int OFFSET = 33;

    @Override
    protected void map(Object key, SequencedFragment value, Context context) throws IOException, InterruptedException {
        final Text quality = value.getQuality();
        for (int i = 0; i < quality.getLength(); i++) {
            context.write(new IntWritable(i + 1), new QualityCountHelperWritable(getCorrespondingIntValue(quality.charAt(i))));
        }
    }

    private int getCorrespondingIntValue(int qualVal) {
        return qualVal - FastQ_PerBaseQual_Mapper.OFFSET;
    }

}